libraries/core/lib/search/objective/BranchObjectiveFunction.ts

```diff
@@ -66,95 +60,41 @@ export class BranchObjectiveFunction<T extends Encoding>
       return Number.MAX_VALUE;
     }
 
-    // let's check if the line is covered
-    if (executionResult.coversLine(this._line)) {
-      const branchTrace = executionResult
-        .getTraces()
-        .find(
-          (trace) =>
-            trace.type === "branch" &&
-            trace.line === this._line &&
-            trace.branchType === this._type
-        );
-
-      if (branchTrace.hits > 0) {
-        return 0;
-      } else {
-        const oppositeBranch = executionResult.getTraces().find(
-          (trace) =>
-            trace.type === "branch" &&
-            trace.id === branchTrace.id && // Same branch id
-            trace.branchType !== this._type // The opposite branch type
-        );
-
-        return BranchDistance.branchDistanceNumeric(
-          oppositeBranch.opcode,
-          oppositeBranch.left,
-          oppositeBranch.right,
-          this._type
-        );
-      }
+    // let's check if the node is covered
+    if (executionResult.coversId(this._id)) {
+      return 0;
     }
 
-    // find the corresponding branch node inside the cfg
-    const branchNode = this._subject.cfg.findNodeOfTypeByLine(
-      this._line,
-      NodeType.Branch
-    );
+    // find the corresponding node inside the cfg
+    const targetNode = this._subject.cfg.graph.getNodeById(this._id);
 
-    // TODO maybe childNode is not required.
-    const childEdge = this._subject.cfg.edges.find((edge) => {
-      return edge.from === branchNode.id && edge.branchType === this._type;
-    });
-    const childNode = this._subject.cfg.nodes.find((node) => {
-      return node.id === childEdge.to;
-    });
+    if (!targetNode) {
+      throw new Error(shouldNeverHappen("BranchObjectiveFunction"));
+    }
 
     // Find approach level and ancestor based on node and covered nodes
     const { approachLevel, closestCoveredBranchTrace } =
-      ApproachLevel.calculate(
-        this._subject.cfg,
-        childNode,
+      this.approachLevel.calculate(
+        this._subject.cfg.graph,
+        targetNode,
         executionResult.getTraces()
       );
 
-    // if closer node (branch or probe) is not found, we return the distance to the root branch
+    // if closest covered node is not found, we return the distance to the root branch
     if (!closestCoveredBranchTrace) {
-      return Number.MAX_VALUE;
+      throw new Error(shouldNeverHappen("BranchObjectiveFunction"));
     }
 
-    let branchDistance: number;
-    if (closestCoveredBranchTrace.type === "function") branchDistance = 1;
-    else branchDistance = this.computeBranchDistance(closestCoveredBranchTrace);
+    const branchDistance = this.branchDistance.calculate(
```

Comment:
We should probably do the normalization in this class as the branch distance itself it not normalized. It is only normalized within the context of combining the two.

Reply:
From how I understand branch distance you should normalize after each operation.
So the normalization would happen ad multiple steps
For example if you have

```a == b && c == d```
You get
```normalize(normalize(abs(a-b)) + normalize(abs(c-d)))```

Reply:
Here is a small example:
The difference in both scenarios is 1.
However, in the first scenario, only one variable has to be changed by 1, while in the other scenario, two variables have to be changed by 0,5.
With the normalization at every step we get 0,33 in the first scenario and 0,4 in the second scenario. This makes sense to me since in the first scenario we only have to change one variable so it is closer.
Without the normalization we get that both scenarios are 0,5.

```
a = 1
b = 1
c = 1
d = 2
```
```
0+1/2=0.5
=> 1,7375/2,7375=0,33
```
Without the normalization before:
```
0+1=1
=> 1/2 = 0,5
```

```
a = 0.5
b = 1
c = 0.5
d = 1
```
```
0,5/1,5+0,5/1,5=1/1,5
=> 0,66/1,66=0,4
```
Without the normalization before:
```
0.5+0.5=1
=> 1/2 = 0,5
```